File: mobile/ios/src/data/facebook-session-provider.swift
import Foundation
import FBSDKLoginKit
import Hydra

class FacebookSessionProvider: SessionProvider {
  fileprivate let config: Config!

  init(config: Config!) {
    self.config = config
  }

  func initialise() -> Promise<Void> {
    return Promise(resolved: ())
  }

  var isAuthenticated: Bool {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return false
      }

      guard let token = AccessToken.current else {
        return false
      }

      return !token.isExpired
    }
  }

  var isExpired: Bool {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return true
      }

      guard let token = AccessToken.current else {
        return false
      }

      return token.isExpired
    }
  }

  var authToken: String? {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return nil
      }

      guard let token = AccessToken.current else {
        return nil
      }

      return token.tokenString
    }
  }

  var accountDescription: String? {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return nil
      }

      let ret = [Profile.current?.firstName, Profile.current?.lastName].compactMap { $0 }.joined(separator: " ")

      if ret.isEmpty {
        return nil
      }

      return ret
    }
  }

  var firstName: String? {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return nil
      }

      return Profile.current?.firstName
    }
  }

  var lastName: String? {
    get {
      guard config.flags.flags.facebookAuth == true else {
        return nil
      }

      return Profile.current?.lastName
    }
  }

  func registerBespokeSession(_ session: Any?) {}
}

File: mobile/android/app/src/main/java/io/meshtech/koasta/data/internal/FacebookSessionProvider.kt
package io.meshtech.koasta.data.internal

import com.facebook.AccessToken
import com.facebook.Profile

class FacebookSessionProvider: IFacebookSessionProvider {
  override fun initialise() {}

  override fun registerBespokeSession(session: Any?) {}

  override val isAuthenticated: Boolean
    get() {
      val token = AccessToken.getCurrentAccessToken() ?: return false
      return !token.isExpired
    }

  override val isExpired: Boolean
    get() {
      val token = AccessToken.getCurrentAccessToken() ?: return false
      return token.isExpired
    }

  override val authToken: String?
    get() {
      val token = AccessToken.getCurrentAccessToken() ?: return null
      return token.token
    }

  override val accountDescription: String?
    get() {
      val profile = Profile.getCurrentProfile() ?: return null

      val firstName = profile.firstName ?: ""
      val lastName = profile.lastName ?: ""

      val ret = listOf(firstName, lastName).joinToString(" ").trim()

      return if (ret.isEmpty()) { null } else ret
    }

  override val firstName: String?
    get() {
      val profile = Profile.getCurrentProfile() ?: return null
      return profile.firstName ?: "Anonymous"
    }

  override val lastName: String?
    get() {
      val profile = Profile.getCurrentProfile() ?: return null
      return profile.lastName ?: ""
    }
}
